### DISCUSSION TITLE
Using directus SSO with AppGyver

### DISCUSSION THREAD

**Discussion by jvictorfranca** [directus/directus#11250] [Archive]
Hi!
 I am currently trying to make a low code application using Directus as backend and Appgyver as frontEnd. It is currently working with the default authentication, but i'm trying to implement a SSO option.  I managed to make the microsoft SSO work, but i am having issues in implementing the microsoft SSO on the appgyver. The thing is that when I try to get the SSO with the GET request at /auth/login/microsoft , the answer seems to be a HTML, not a JSON. But appgyver does not seem to read the HTML answer, the closest i could do was open the link (localhost:8055/auth/login/microsoft) in appgyver but in other tab, and i cant retrieve the token after i login. So i was thinking if there was a way to login in with microsoft SSO without a browser, or a way to put the token in the other tab inside appgyver. Any insights?

Thanks a lot! .

**Comment by jamescammarano** [comment]
Hi!
This sounds like more of an issue with appgyver if I'm understanding correctly. Let me know if I'm misunderstanding though!

**Comment by jvictorfranca** [comment]
Mostly yeah, an appgyver issue. But i wanted to know if there was another way to use the SSO instead of the docs GET request  Because using MSAL its possible to login without a browser, but as directus doest give access to MSAL directly i was wondering if that was possible

**Comment by rijkvanzanten** [comment]
@jvictorfranca I'm not familiar with AppGyver or how it works.. Would you mind walking me through the scenario here? You're putting together a website/app in a page-builder from what it looks like, and you're trying to log into Directus from this app?

**Reply by jvictorfranca** [reply]
Exactly. Appgyver is a low code frontend. So it has its modules for making the frontend. I was using Directus as backend and default authentication to authenticate the appgyver application, and it was working fine (appgyver only have a form type structure for making requests, so you cannot work with the response as easy as in code, only can use JSON as response for what ive tested so far). As the default Directus user login is only a post request to Directus (POST http://localhost:8055/auth/login), with the response being a JSON with the auth token, appgyver could handle it. Now im trying to implement a SSO  microsoft authentication. I could implement it in Directus, and it is working for login in Directus with the SSO, but when i checked in postman, the response to the get request for the SSO  (GET http://localhost:8055/auth/login/microsoft)  is a HTML, which appgyver request cannot read (Claiming CORS error, but any other request from directus works, so i thinks its not CORS at all). So the only way to authenticate via SSO is opening the link (http://localhost:8055/auth/login/microsoft) in a browser, and getting the response token via window.json or localstorage after the microsoft login. In the end all appgyver only allows me to use JSON responses, or opening links in other tabs (the latter i tried to use, but i cannot access the other tab data in appgyver enviroment,  so i cannot put the token inside appgyver). 
So in short im trying to find a way to get the authentication token of Directus via microsoft SSO without using a browser,only with API requests. Im starting to think its not possible for appgyver to support the SSO option, since it does not have a popup module for opening the microsoft login page,  but im kinda lost (and they're not responding anyways haha).
Thats it, thanks so much for the help, i appreciated it, even though i think theres not much i can do, but thanks a lot <3.

**Reply by aidenfoxx** [reply]
The URL http://localhost:8055/auth/login/microsoft is expected to return HTML as it is designed to redirect the user to the SSO login page. I believe what you want to do is send your user to "http://localhost:8055/auth/login/microsoft?redirect=http://yourreturndomain.com", allow them to authenticate, and when they return they will have an auth cookie in their session?

Bear in mind this approach will be restricted to a single domain origin e.g. "yourreturndomain.com" and "*.yourreturndomain.com" and may require setting SameSite="None" and "Secure" in your Directus cookie settings. https://docs.directus.io/configuration/sso/

EDIT: Missed the part about MSAL. We currently don't support MSAL over the login URL, but I believe you could implement it if you want to submit a PR.

**Reply by jvictorfranca** [reply]
Thanks so much! That was what i wanted for redirecting, but rn i am having an issue that the cookies arent being set. Im trying to fix it with this discussion https://github.com/directus/directus/discussions/9720 (i think its because im running everything locally, tried to change localhost for http://127.0.0.1:8055/ and using firefox, but no success). Should I open another question thread if I fail?
Thanks :)

**Reply by aidenfoxx** [reply]
Make sure your service and directus are running on the same domain (and I believe same port). If you're using "Same site=None; Secure" cookies you need to be running on https. Firefox will probably be required for testing in localhost.

**Reply by jvictorfranca** [reply]
I tried even in the same port but its still not working. The setcookie header and the cookie is comming from login, but it does not sticks in the  domain page:

![image](https://user-images.githubusercontent.com/83782983/151167748-e9eff56a-6d97-4657-9c71-c52f02983c89.png)

![image](https://user-images.githubusercontent.com/83782983/151167832-a05a4c8a-7e0f-4468-94a6-062fbf88dd0d.png)


I tried also with firefox and secure false, or same_site: lax, neither worked T.T

**Reply by aidenfoxx** [reply]
"SameSite=None" also requires "Secure" to be set, which in turn requires you to be on "https".

**Reply by jvictorfranca** [reply]
So, theres no way to test it in localhost? I would have to get a ssh certificate?

**Reply by aidenfoxx** [reply]
You can run https locally with a self signed certificate. It's not hard to do, but you'll have to find a guide on it elsewhere.

**Reply by jvictorfranca** [reply]
(sorry for this, im totally lost haha). I got a node application running on HTTPS, but i cant change directus default server outside of http://localhost as its server is in source code. So one option would be run directus on ngnix as discussed in https://github.com/directus/directus/discussions/3307, then i would have a https access and would be able to set cookies right? 

And just to be sure, the only ways to get the tokens with the SSO are by the GET request (https://docs.directus.io/reference/authentication/#login-using-sso-providers) that i cannot use; and by seamless SSO with cookies (https://docs.directus.io/configuration/sso/#seamless-sso), which i cannot do locally because of the http? 

Thanks again, and sorry for the ignorance haha

**Reply by aidenfoxx** [reply]
Huh... I actually didn't realize directus couldn't be run under SSL. Either way, that can be easily solved by proxying Directus through express if you don't want to install nginx.

But really if you're planning on running your client and directus instance on the same URL and port e.g. "http://localhost:8055", then you don't need "SameSite=None" and "Secure". That's only required for wildcards and other sub-domains (I believe). It should be sufficient to use "SameSite=Lax".

Cookies are a pain in the ass these days. Unfortunately the price we pay for security. 🤷

**Reply by jvictorfranca** [reply]
I see. If i can use "SameSite=Lax" and Secure = false my cookie set error is still happening T.T , the same this attempt to set a cookie with Set-Cookie header was blocked because its Domain attribute was invalid with regards to the current host url.

My full env in case im doing something wrong

## General

PORT=8055
PUBLIC_URL="http://localhost:8055"

####################################################################################################
## Database

DB_CLIENT="mssql"
DB_HOST="127.0.0.1"
DB_PORT="1433"
DB_DATABASE="directusTest"
DB_USER="sa"
DB_PASSWORD="password"
DB_OPTIONS__ENCRYPT="false"


####################################################################################################
## Rate Limiting

RATE_LIMITER_ENABLED=false
RATE_LIMITER_STORE=memory
RATE_LIMITER_POINTS=25
RATE_LIMITER_DURATION=1

####################################################################################################
## Cache

CACHE_ENABLED=false

####################################################################################################
## File Storage

STORAGE_LOCATIONS="local"
STORAGE_LOCAL_DRIVER="local"
STORAGE_LOCAL_ROOT="./uploads"

####################################################################################################
## Security

KEY="myKey"
SECRET="mySecret"


ACCESS_TOKEN_TTL="15m"
REFRESH_TOKEN_TTL="7d"
REFRESH_TOKEN_COOKIE_SECURE=false
REFRESH_TOKEN_COOKIE_SAME_SITE="lax"
REFRESH_TOKEN_COOKIE_NAME="directus_refresh_token"

####################################################################################################
## Auth Providers

AUTH_PROVIDERS=''

####################################################################################################
## Extensions

EXTENSIONS_PATH="./extensions"

####################################################################################################
## Email

EMAIL_FROM="no-reply@directus.io"
EMAIL_TRANSPORT="sendmail"
EMAIL_SENDMAIL_NEW_LINE="unix"
EMAIL_SENDMAIL_PATH="/usr/sbin/sendmail"

#######################################################################################################


########################################################################################################

AUTH_PROVIDERS="microsoft"
AUTH_DISABLE_DEFAULT ="false"
AUTH_MICROSOFT_DRIVER="openid"
AUTH_MICROSOFT_CLIENT_ID="myClientId"
AUTH_MICROSOFT_CLIENT_SECRET="mySecret"
AUTH_MICROSOFT_SCOPE="openid profile email"
AUTH_MICROSOFT_IDENTIFIER_KEY="email"
AUTH_MICROSOFT_ICON="microsoft"
AUTH_MICROSOFT_ISSUER_URL="https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration"

#############################################################################################################

REFRESH_TOKEN_COOKIE_DOMAIN="http://localhost:8055"
REFRESH_TOKEN_COOKIE_SECURE="false"
REFRESH_TOKEN_COOKIE_SAME_SITE="lax"



I am trying to authenticate at localhost:8055/auth/login/microsoft?redirect=http://localhost:8055/tes/test, that is a directus custom endpoint, but tried also to redirect to localhost:3000, in both cases with the same error:
Cookie gets to localhost:8055/auth/login/microsoft/callback , setcookies in the header, but the cookie itself does not get to the last url (i tink cookie_domain, that i tried localhost3000 and 8055). 

Thanks once again :)

**Reply by jvictorfranca** [reply]
EDIT: Now i was able to get the refresh token to localhost:3000 , what i did was take off the REFRESH_TOKEN_COOKIE_DOMAIN. Now I need to redirect to https://platform.preview.appgyver.com/, but im getting the set-Cookie same error again. I put back the DOMAIN and tried with securae, samesite = none and lax and couldnt make it. Any ideas? Thanks <3

**Reply by aidenfoxx** [reply]
I thought I made it clear early on, but cross-domain cookies won't work here. Your app and Directus instance need to be on the same domain (or wildcard domain).

**Comment by jvictorfranca** [comment]
EDIT: I see now there's a error in the set-cookie header: this attempt to set a cookie with Set-Cookie header was blocked because its Domain attribute was invalid with regards to the current host url, even when i use the same domain (http://localhost:8055) with or without secure.
